src/main/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListBatchValidator.java

```diff
@@ -0,0 +1,222 @@
+        + "with key '%s' is null.";
+    static final String TEXT_LINE_ITEM_DRAFT_NAME_NOT_SET = "LineItemDraft at position '%d' of "
+        + "ShoppingListDraft with key '%s' has no name set. Please make sure all textLineItems have names.";
+
+    public ShoppingListBatchValidator(@Nonnull final ShoppingListSyncOptions syncOptions,
+                                      @Nonnull final ShoppingListSyncStatistics syncStatistics) {
+        super(syncOptions, syncStatistics);
+    }
+
+    /**
+     * Given the {@link List}&lt;{@link ShoppingListDraft}&gt; of drafts this method attempts to validate
+     * drafts and collect referenced type keys from the draft and return an {@link ImmutablePair}&lt;{@link Set}&lt;
+     * {@link ShoppingListDraft}&gt;,{@link ReferencedKeys}&gt; which contains the {@link Set} of valid drafts and
+     * referenced keys.
+     *
+     * <p>A valid shopping list draft is one which satisfies the following conditions:
+     * <ol>
+     * <li>It is not null</li>
+     * <li>It has a key which is not blank (null/empty)</li>
+     * <li>It has a name which is not null</li>
+     * <li>It has all lineItems AND textLineItems valid</li>
+     * <li>A lineItem is valid if it satisfies the following conditions:
+     * <ol>
+     * <li>It has a SKU which is not blank (null/empty)</li>
+     * </ol>
+     * <li>A textLineItem is valid if it satisfies the following conditions:
+     * <ol>
+     * <li>It has a name which is not blank (null/empty)</li>
+     * </ol>
+     * </li>
+     * </ol>
+     *
+     * @param shoppingListDrafts the shopping list drafts to validate and collect referenced type keys.
+     * @return {@link ImmutablePair}&lt;{@link Set}&lt;{@link ShoppingListDraft}&gt;,
+     * {@link ReferencedKeys}&gt; which contains the {@link Set} of valid drafts and
+     *      referenced keys within a wrapper.
+     */
+    @Override
+    public ImmutablePair<Set<ShoppingListDraft>, ReferencedKeys> validateAndCollectReferencedKeys(
+        @Nonnull final List<ShoppingListDraft> shoppingListDrafts) {
+        final ReferencedKeys referencedKeys = new ReferencedKeys();
+        final Set<ShoppingListDraft> validDrafts = shoppingListDrafts
+            .stream()
+            .filter(this::isValidShoppingListDraft)
+            .peek(shoppingListDraft ->
+                collectReferencedKeys(referencedKeys, shoppingListDraft))
+            .collect(Collectors.toSet());
+
+        return ImmutablePair.of(validDrafts, referencedKeys);
+    }
+
+    private boolean isValidShoppingListDraft(
+        @Nullable final ShoppingListDraft shoppingListDraft) {
+
+        if (shoppingListDraft == null) {
+            handleError(SHOPPING_LIST_DRAFT_IS_NULL);
+        } else if (isBlank(shoppingListDraft.getKey())) {
+            handleError(format(SHOPPING_LIST_DRAFT_KEY_NOT_SET, shoppingListDraft.getName()));
+        } else if (shoppingListDraft.getName() == null || shoppingListDraft.getName().getLocales().isEmpty()) {
```

Comment:
validation of the `LocalisedString` could be a method, so we don't need to duplicate code in line: 152(getTextLineItemDraftErrorsInAllTextLineItems)

Reply:
36d2868